File: model-client/src/main/java/org/modelix/model/api/IBranch.java
package org.modelix.model.api;

import java.util.function.Supplier;

public interface IBranch {
  void runRead(Runnable runnable);
  <T> T computeRead(Supplier<T> computable);
  void runWrite(Runnable runnable);
  <T> T computeWrite(Supplier<T> computable);

  boolean canRead();
  boolean canWrite();

  ITransaction getTransaction();
  IReadTransaction getReadTransaction();
  IWriteTransaction getWriteTransaction();

  void addListener(IBranchListener l);
  void removeListener(IBranchListener l);
}

File: model-client/src/main/kotlin/org/modelix/model/api/IBranch.kt
package org.modelix.model.api

import java.util.function.Supplier

interface IBranch {
    fun runRead(runnable: Runnable?)
    fun <T> computeRead(computable: Supplier<T>?): T
    fun runWrite(runnable: Runnable?)
    fun <T> computeWrite(computable: Supplier<T>?): T
    fun canRead(): Boolean
    fun canWrite(): Boolean
    val transaction: ITransaction?
    val readTransaction: IReadTransaction?
    val writeTransaction: IWriteTransaction?
    fun addListener(l: IBranchListener?)
    fun removeListener(l: IBranchListener?)
}
